### ISSUE TITLE
plugins.okru: HLS playlist URL expires

### ISSUE THREAD

**Issue by leomani51214** [streamlink/streamlink#6369]
### Checklist

- [x] This is a [plugin issue](https://streamlink.github.io/plugins.html) and not [a different kind of issue](https://github.com/streamlink/streamlink/issues/new/choose)
- [x] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [x] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [x] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Collaboration

- [x] [I will provide feedback should a pull request be opened with a fix for the plugin](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#pull-request-feedback)

### Streamlink version

streamlink 7.1.1

### Description

Hello,
I am using the following Streamlink command to extract the live stream URL for a broadcast that I created on ok.ru :
streamlink "https://ok.ru/video/8532062379616" --stream-url
This successfully returns a .m3u8 link for the stream :
https://vsd242.okcdn.ru/hls/7390660069984.m3u8/sig/iPJ8iZErCo/expires/1735602941172/srcIp/39.63.235.143/urls/45.136.20.58/clientType/0/srcAg/GECKO/mid/8532062379616/video.m3u8?p
However, there is an issue with the validity of the URL. Even though I can extract this .m3u8 link, it doesn't play after one minute of extraction. Upon decoding the expires parameter in the URL, I discovered that the link has an expiration date set to July 2024, while the current date is December 2024.

Of course the stream is off now but I tried many times but StreamLink always return wrong URL. 

This is the primary reason why the stream doesn't work as expected.

Question:

Can anyone help me understand why the link isn't working, and provide guidance on how to refresh or extend the expiration time so that it works correctly? Specifically, I need a way to ensure that the stream link remains valid for a longer period of time.

### Debug log

```text
C:\Users\AL MAkkah>streamlink -l debug "https://ok.ru/live/8533722079072" 480p
[session][debug] Loading plugin: okru
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.12.8
[cli][debug] OpenSSL:    OpenSSL 3.0.15 3 Sep 2024
[cli][debug] Streamlink: 7.1.1
[cli][debug] Dependencies:
[cli][debug]  certifi: 2024.12.14
[cli][debug]  isodate: 0.7.2
[cli][debug]  lxml: 5.3.0
[cli][debug]  pycountry: 24.6.1
[cli][debug]  pycryptodome: 3.21.0
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.32.3
[cli][debug]  trio: 0.28.0
[cli][debug]  trio-websocket: 0.11.1
[cli][debug]  urllib3: 2.3.0
[cli][debug]  websocket-client: 1.8.0
[cli][debug] Arguments:
[cli][debug]  url=https://ok.ru/live/8533722079072
[cli][debug]  stream=['480p']
[cli][debug]  --loglevel=debug
[cli][debug]  --ffmpeg-ffmpeg=C:\Users\AL MAkkah\AppData\Local\Programs\Streamlink\ffmpeg\ffmpeg.exe
[cli][info] Found matching plugin okru for URL https://ok.ru/live/8533722079072
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 240p (worst), 360p, 480p, 720p (best)
[cli][info] Opening stream: 480p (hls)
[cli][info] Starting player: C:\Program Files\VideoLAN\VLC\vlc.exe
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 32; Last Sequence: 56
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 54; End Sequence: None
[stream.hls][debug] Adding segment 54 to queue
[stream.hls][debug] Adding segment 55 to queue
[stream.hls][debug] Adding segment 56 to queue
[stream.hls][debug] Writing segment 54 to output
[stream.hls][debug] Segment 54 complete
[cli.output][debug] Opening subprocess: ['C:\\Program Files\\VideoLAN\\VLC\\vlc.exe', '--input-title-format', 'https://ok.ru/live/8533722079072', '-']
[stream.hls][debug] Writing segment 55 to output
[stream.hls][debug] Segment 55 complete
[stream.hls][debug] Writing segment 56 to output
[stream.hls][debug] Segment 56 complete
[cli][debug] Writing stream to output
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 57 to queue
[stream.hls][debug] Writing segment 57 to output
[stream.hls][debug] Segment 57 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 58 to queue
[stream.hls][debug] Adding segment 59 to queue
[stream.hls][debug] Writing segment 58 to output
[stream.hls][debug] Segment 58 complete
[stream.hls][debug] Writing segment 59 to output
[stream.hls][debug] Segment 59 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 60 to queue
[stream.hls][debug] Writing segment 60 to output
[stream.hls][debug] Segment 60 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 61 to queue
[stream.hls][debug] Adding segment 62 to queue
[stream.hls][debug] Writing segment 61 to output
[stream.hls][debug] Segment 61 complete
[stream.hls][debug] Writing segment 62 to output
[stream.hls][debug] Segment 62 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 63 to queue
[stream.hls][debug] Writing segment 63 to output
[stream.hls][debug] Segment 63 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 64 to queue
[stream.hls][debug] Adding segment 65 to queue
[stream.hls][debug] Writing segment 64 to output
[stream.hls][debug] Segment 64 complete
[stream.hls][debug] Writing segment 65 to output
[stream.hls][debug] Segment 65 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 66 to queue
[stream.hls][debug] Writing segment 66 to output
[stream.hls][debug] Segment 66 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 67 to queue
[stream.hls][debug] Adding segment 68 to queue
[stream.hls][debug] Writing segment 67 to output
[stream.hls][debug] Segment 67 complete
[stream.hls][debug] Writing segment 68 to output
[stream.hls][debug] Segment 68 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 69 to queue
[stream.hls][debug] Writing segment 69 to output
[stream.hls][debug] Segment 69 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 70 to queue
[stream.hls][debug] Adding segment 71 to queue
[stream.hls][debug] Writing segment 70 to output
[stream.hls][debug] Segment 70 complete
[stream.hls][debug] Writing segment 71 to output
[stream.hls][debug] Segment 71 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 72 to queue
[stream.hls][debug] Writing segment 72 to output
[stream.hls][debug] Segment 72 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 73 to queue
[stream.hls][debug] Adding segment 74 to queue
[stream.hls][debug] Writing segment 73 to output
[stream.hls][debug] Segment 73 complete
[stream.hls][debug] Writing segment 74 to output
[stream.hls][debug] Segment 74 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 75 to queue
[stream.hls][debug] Writing segment 75 to output
[stream.hls][debug] Segment 75 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 76 to queue
[stream.hls][debug] Adding segment 77 to queue
[stream.hls][debug] Writing segment 76 to output
[stream.hls][debug] Segment 76 complete
[stream.hls][debug] Writing segment 77 to output
[stream.hls][debug] Segment 77 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 78 to queue
[stream.hls][debug] Writing segment 78 to output
[stream.hls][debug] Segment 78 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 79 to queue
[stream.hls][debug] Adding segment 80 to queue
[stream.hls][debug] Writing segment 79 to output
[stream.hls][debug] Segment 79 complete
[stream.hls][debug] Writing segment 80 to output
[stream.hls][debug] Segment 80 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 81 to queue
[stream.hls][debug] Writing segment 81 to output
[stream.hls][debug] Segment 81 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 82 to queue
[stream.hls][debug] Adding segment 83 to queue
[stream.hls][debug] Writing segment 82 to output
[stream.hls][debug] Segment 82 complete
[stream.hls][debug] Writing segment 83 to output
[stream.hls][debug] Segment 83 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 84 to queue
[stream.hls][debug] Writing segment 84 to output
[stream.hls][debug] Segment 84 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 85 to queue
[stream.hls][debug] Adding segment 86 to queue
[stream.hls][debug] Writing segment 85 to output
[stream.hls][debug] Segment 85 complete
[stream.hls][debug] Writing segment 86 to output
[stream.hls][debug] Segment 86 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 87 to queue
[stream.hls][debug] Writing segment 87 to output
[stream.hls][debug] Segment 87 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 88 to queue
[stream.hls][debug] Adding segment 89 to queue
[stream.hls][debug] Writing segment 88 to output
[stream.hls][debug] Segment 88 complete
[stream.hls][debug] Writing segment 89 to output
[stream.hls][debug] Segment 89 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 90 to queue
[stream.hls][debug] Writing segment 90 to output
[stream.hls][debug] Segment 90 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 91 to queue
[stream.hls][debug] Adding segment 92 to queue
[stream.hls][debug] Writing segment 91 to output
[stream.hls][debug] Segment 91 complete
[stream.hls][debug] Writing segment 92 to output
[stream.hls][debug] Segment 92 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 93 to queue
[stream.hls][debug] Writing segment 93 to output
[stream.hls][debug] Segment 93 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 94 to queue
[stream.hls][debug] Adding segment 95 to queue
[stream.hls][debug] Writing segment 94 to output
[stream.hls][debug] Segment 94 complete
[stream.hls][debug] Writing segment 95 to output
[stream.hls][debug] Segment 95 complete
[stream.hls][debug] Reloading playlist
[stream.hls][warning] Failed to reload playlist: Unable to open URL: https://vsd94.okcdn.ru/hls/7392547113568.m3u8/sig/CFn76cXGfcA/expires/1735637667920/srcIp/39.63.251.103/urls/45.136.21.45/clientType/0/srcAg/GECKO/mid/8533722079072/7392547113568_medium/index.m3u8 (404 Client Error: Not Found for url: https://vsd94.okcdn.ru/hls/7392547113568.m3u8/sig/CFn76cXGfcA/expires/1735637667920/srcIp/39.63.251.103/urls/45.136.21.45/clientType/0/srcAg/GECKO/mid/8533722079072/7392547113568_medium/index.m3u8)
[stream.hls][debug] Reloading playlist
[stream.hls][warning] Failed to reload playlist: Unable to open URL: https://vsd94.okcdn.ru/hls/7392547113568.m3u8/sig/CFn76cXGfcA/expires/1735637667920/srcIp/39.63.251.103/urls/45.136.21.45/clientType/0/srcAg/GECKO/mid/8533722079072/7392547113568_medium/index.m3u8 (404 Client Error: Not Found for url: https://vsd94.okcdn.ru/hls/7392547113568.m3u8/sig/CFn76cXGfcA/expires/1735637667920/srcIp/39.63.251.103/urls/45.136.21.45/clientType/0/srcAg/GECKO/mid/8533722079072/7392547113568_medium/index.m3u8)
[stream.hls][debug] Reloading playlist
[stream.hls][warning] Failed to reload playlist: Unable to open URL: https://vsd94.okcdn.ru/hls/7392547113568.m3u8/sig/CFn76cXGfcA/expires/1735637667920/srcIp/39.63.251.103/urls/45.136.21.45/clientType/0/srcAg/GECKO/mid/8533722079072/7392547113568_medium/index.m3u8 (404 Client Error: Not Found for url: https://vsd94.okcdn.ru/hls/7392547113568.m3u8/sig/CFn76cXGfcA/expires/1735637667920/srcIp/39.63.251.103/urls/45.136.21.45/clientType/0/srcAg/GECKO/mid/8533722079072/7392547113568_medium/index.m3u8)
[stream.hls][warning] No new segments in playlist for more than 9.00s. Stopping...
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

**Comment by leomani51214** [comment]
The bebug, best command work perfectly for a stream the video gets opened in the VLC but my problem is that I need the direct stream link m3u8 of a live stream. The one I get using  --stream-url is not working.

**Comment by bastimeyer** [comment]
> best command work perfectly for a stream

No, it does not, as you can clearly see in the debug log output. The HLS playlist URL expires and needs to be re-acquired somehow.

However, I've been running some streams for some time now, and none of them have expired so far. The expiration parameter in the stream URL has plenty of head room, so I'm really not sure yet why some other streams expire. Do they even expire or are they just going offline? A 404 HTTP error (not found) implies that the resource is not available anymore. A 401 (unauthorized) or 403 (forbidden) error would make much more sense for an expired and now inaccessible resource.

> The one I get using --stream-url is not working

There's no "--stream-url one" or "Streamlink-internal one". The plugin returns the same kinds of HLS playlist URLs. In the first case, the resolved HLS playlist URL is simply written to stdout, and in the second case it's used in Streamlink's own HLS client implementation that then also launches the video player in its default configuration.

If you can't access the HLS playlist URL in a different program, then this means that you're missing certain HTTP headers which are set by default on Streamlink's HTTP session, like the user agent header for example, or HTTP headers which are configured by the used plugin, of which there are none in case of okru.

```
$ curl -sI $(streamlink --stream-url https://ok.ru/live/9537408409327)
HTTP/1.1 403 Forbidden
Content-Length: 0
Access-Control-Allow-Origin: *
Expires: Mon, 30 Dec 2024 16:43:26 GMT
Connection: Keep-Alive

$ curl -sI -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:133.0) Gecko/20100101 Firefox/133.0' $(streamlink --stream-url https://ok.ru/live/9537408409327)
HTTP/1.1 200 OK
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
Access-Control-Allow-Headers: Origin, X-Requested-With, Accept, Range
Access-Control-Expose-Headers: Date
Access-Control-Expose-Headers: X-Playback-Duration, X-Playback-Duration-Millis
date: Mon, 30 Dec 2024 12:43:46 GMT
content-length: 468
expires: Mon, 30 Dec 2024 12:43:48 GMT
content-type: application/vnd.apple.mpegurl;
connection: Keep-Alive
cache-control: must-revalidate
X-Playback-Duration: 7200000
```

**Comment by leomani51214** [comment]
I think I might not have explained the issue clearly earlier. Let me try again.

I start a stream on OK.ru and then I run the following command:

C:\Users\Streamer>streamlink -l debug "https://ok.ru/live/8536079146848" --stream-url
https://vsd111.okcdn.ru/hls/7395869133408.m3u8/sig/tEvumw73HIs/expires/1735677278825/srcIp/39.63.235.143/urls/45.136.20.58/clientType/0/srcAg/GECKO/mid/8536079146848/video.m3u8?p

The issue is that the link I get from this command is not working.

Of course, when you try to test this, the stream will likely be offline because I don’t know when you’ll run it. However, I need help figuring out why the link doesn’t work even when the stream is live.

**Comment by bastimeyer** [comment]
> The issue is that the link I get from this command is not working.

Read the second part of my comment. You are missing HTTP headers.

**Comment by leomani51214** [comment]
sorry bro, but I am very new in this stream world and I didn't understand what you meant can you please explain it a bit more and if with an example then it will be great. Thank You!!!

**Comment by bastimeyer** [comment]
Sorry, but I'm not going to explain basic HTTP functionality to you. If you don't know how HTTP works, then please use Google. The URL alone is not enough to access the HLS playlist. Once again, you also need the `User-Agent` header. My `curl` example shows this very clearly.

This is btw offtopic here on this bug tracker. This thread is about expiration times. And regarding that, I just let some live streams play for over an hour, without any (early) expirations. As said, a 404 error implies that the stream ended on their end. Also not sure what you mean with 2 seconds. The expiration time that's part of the URL is 24 hours, which is acceptable for Streamlink's plugins.

```sh
expires=$(streamlink --stream-url https://ok.ru/live/9537408409327 | grep -oE 'expires/[^/]+' | cut -d/ -f2)
now=$(date -u +%s)
echo $(( ( $expires - $now * 1000 ) / 1000 ))
```
```
86400
```

I'm going to close this issue now, since I believe that there is no **early** expiration issue.

**Comment by leomani51214** [comment]
I understood your earlier curl example and extracting the streaming url by using that method but the link I get from the command is still not working.

The Command I run
C:\Users\Stream>for /f "delims=" %A in ('streamlink --stream-url  https://ok.ru/live/9537408409327') do curl -sI -H "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:133.0) Gecko/20100101 Firefox/133.0" "%A"

and the response I get
C:\Users\Stream>curl -sI -H "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:133.0) Gecko/20100101 Firefox/133.0" "https://vsd161.okcdn.ru/hls/7892687063791.m3u8/sig/R5kqI_ZMpPM/expires/1735682371221/srcIp/39.63.235.143/urls/185.226.55.74/clientType/0/srcAg/GECKO/mid/9537408409327/video.m3u8?p"
HTTP/1.1 200 OK
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
Access-Control-Allow-Headers: Origin, X-Requested-With, Accept, Range
Access-Control-Expose-Headers: Date
Access-Control-Expose-Headers: X-Playback-Duration, X-Playback-Duration-Millis
date: Mon, 30 Dec 2024 21:59:33 GMT
content-length: 468
expires: Mon, 30 Dec 2024 21:59:35 GMT
content-type: application/vnd.apple.mpegurl;
connection: Keep-Alive
cache-control: must-revalidate
X-Playback-Duration: 7200000

But the problem is still same and the .m3u8 link that I am getting is not playing.
